modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/resource/FDSHeadlessResourcesUtil.java

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.data.set.views.web.internal.resource;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.Dictionary;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.Filter;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * @author Marko Cikos
+ */
+@Component(service = {})
+public class FDSHeadlessResourcesUtil {
+
+	public static List<FDSHeadlessResource> getFDSHeadlessResources() {
+		return new ArrayList<>(_fdsHeadlessResourcesMap.values());
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext)
+		throws InvalidSyntaxException {
+
+		Filter filter = bundleContext.createFilter(
+			"(osgi.jaxrs.resource=true)");
+
+		_serviceTracker = new ServiceTracker<>(
+			bundleContext, filter,
+			new FDSHeadlessResourceServiceTrackerCustomizer(bundleContext));
+
+		_serviceTracker.open();
+	}
+
+	private static final Map<String, FDSHeadlessResource>
+		_fdsHeadlessResourcesMap = new ConcurrentHashMap<>();
+	private static ServiceTracker<Object, Object> _serviceTracker;
+
+	private class FDSHeadlessResourceServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer<Object, Object> {
+
+		@Override
+		public Object addingService(ServiceReference<Object> serviceReference) {
+			String entityClassName = (String)serviceReference.getProperty(
+				"entity.class.name");
+
+			if (entityClassName != null) {
+				String[] entityClassNameParts = StringUtil.split(
+					entityClassName, StringPool.PERIOD);
+
+				_fdsHeadlessResourcesMap.put(
```

Comment:
For Objects, I don't know if the parsing is what you expected:

![image](https://user-images.githubusercontent.com/32699538/217505183-6483e874-60d1-4096-a5f5-b6cea5fdec00.png)

Reply:
Hmm this is a product and a design call. We probably want to highlight that something is an `Object`, in a unique way. I think we can leave this for now as it is. Display of items is temporary anyway, see https://github.com/liferay-frontend/liferay-portal/pull/3031#discussion_r1099285775

Reply:
btw, it was a pleasant surprise to me today that Objects support already works :sweat_smile: It didn't for me last couple of days because I had an issue with stale DB. Fresh DB after `ant all` did the trick.

Reply:
> btw, it was a pleasant surprise to me today that Objects support already works sweat_smile It didn't for me last couple of days because I had an issue with stale DB. Fresh DB after `ant all` did the trick.

Get used to it every day :trollface:

Reply:
We probably need to make some branching logic for the case of objects here, this can be done in a separate ticket but we'll have to consider the case for sure, unless we change the way we fetch the information....

@ccorreagg @LuismiBarcos what if we take this from the [place where we generate](https://github.com/liferay/liferay-portal/blob/55ab82d5372e328c1597e2a1e62eea4f036526d6/modules/apps/headless/headless-discovery/headless-discovery-impl/src/main/java/com/liferay/headless/discovery/internal/jaxrs/resource/v1_0/HeadlessDiscoveryOpenAPIResourceImpl.java#L253-L295) /o/api listing? looks like a good fit for objects and takes into account companyId scoped apis right?

Reply:
Maybe I am misunderstanding the feature 😅 
The purpose is to be able to select an Entity (a blog, a web content, an Object...) and return all the elements of that entity (all blogs, for example). Is that correct?

Or do you just want to list the Entity fields?

For the second case, I think there is no problem but could be tricky as you need the final OpenAPI and get the information from there.

For the first case... as we said, AFAIK we cannot ensure that an API has an endpoint to retrieve all the entities, at least for now. Maybe we need to synchronize to see the requirements and provide a solution.

Reply:
These are two approaches that achieve the same thing, resources list and their fields. Approach in PR goes with backend, leveraging OSGi framework. As an alternative, we could take info just from OpenAPI framework. The end data is the same.

I feel that backend approach offers more flexibility. For example, with https://github.com/liferay-frontend/liferay-portal/pull/3031#issuecomment-1422954692 in mind, we may need a filter on a new custom property. I believe this can only be done in backend.

Another argument for backend is that we can efficiently prepare a flat list of all resources. With OpenAPI, we need to query for all apps, and then query once for each resource within an app.

Reply:
@LuismiBarcos the purpose at this moment is to access the REST application  metadata, so that the dataset application can offer a UI to map the schema fields offered by the app entities to visual elements like table columns or card title, etc. Then, more advanced things like sorting and filtering could be configured from the dataset app. In the end, we will be able to provide a full-fledged configuration to the Frontend Dataset component, which includes the endpoint URLs and some other stuff.

As we make progress, we realized that not every exposed API may include all the necessary things, hence the need of filtering the set of APIs we want to show to the administrator.

Backend approach is probably much better in terms of introspection capabilities, will give us more chances to tweak the way we select those rest apps and its endpoints.

Still, in doing so, the way we process that info in the `/o/api` listing looks very interesting if we want to offer a consistent experience in the dataset app, so probably it worths keeping an eye on it.